trieste/models/gpflow/models.py

```diff
@@ -507,13 +496,308 @@ def conditional_predict_y(
+                f" shape {self.model.inducing_variable.Z.shape} of existing query points."
+                f" Trailing dimensions must match."
+            )
+
+        if dataset.observations.shape[-1] != q_mu.shape[-1]:
+            raise ValueError(
+                f"Shape {dataset.observations.shape} of new observations is incompatible with"
+                f" shape {self.model.q_mu.shape} of existing observations. Trailing"
+                f" dimensions must match."
+            )
+
+        num_data = dataset.query_points.shape[0]
+        self.model.num_data.assign(num_data)
+        self.update_posterior_cache()
+
+        if self._inducing_point_selector is not None:
+            new_inducing_points = self._inducing_point_selector.calculate_inducing_points(
+                current_inducing_points, self, dataset
+            )
+            if not tf.reduce_all(
+                tf.math.equal(
+                    new_inducing_points,
+                    current_inducing_points,
+                )
+            ):  # only bother updating if points actually change
+                self._update_inducing_variables(new_inducing_points)
+
+    def _update_inducing_variables(self, new_inducing_points: TensorType) -> None:
+        """
+        When updating the inducing points of a model, we must also update the other
+        inducing variables, i.e. `q_mu` and `q_sqrt` accordingly. The exact form of this update
+        depends if we are using whitened representations of the inducing variables.
+        See :meth:`_whiten_points` for details.
+
+        :param new_inducing_points: The desired values for the new inducing points.
+        :raise NotImplementedError: If we try to update the inducing variables of a model
+            that has more than one set of inducing points.
+        """
+
+        if isinstance(new_inducing_points, list):
+            raise NotImplementedError(
+                f"""
+                We do not currently support updating models with multiple sets of
+                inducing points however received; {new_inducing_points}
+                """
+            )
+
+        old_inducing_points, _, _, _ = self.get_inducing_variables()
+        tf.assert_equal(
+            tf.shape(old_inducing_points), tf.shape(new_inducing_points)
+        )  # number of inducing points must not change
+
+        if isinstance(self.model.inducing_variable, SharedIndependentInducingVariables):
+            # gpflow says inducing_variable might be a ndarray; it won't
+            cast(TensorType, self.model.inducing_variable.inducing_variable).Z.assign(
+                new_inducing_points
+            )  # [M, D]
+        else:
+            self.model.inducing_variable.Z.assign(new_inducing_points)  # [M, D]
+
+    def get_inducing_variables(
+        self,
+    ) -> Tuple[Union[TensorType, list[TensorType]], TensorType, TensorType, bool]:
```

Comment:
this return type for `SupportsGetInducingVariables` makes sense for SVGP and VGP but not really for SGPR, should we create another interface for q objects?
@henrymoss and @uri-granta any thoughts?

Reply:
Can you clarify? Should SGPR models work with any rule that requires SupportsGetInducingVariables models? (At the moment that's just DecoupledTrajectorySampler I think, and that already supports SGPR as a FeatureDecompositionInternalDataModel.) If not, what is actually going to call get_inducing_variables on an SGPR?